File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> res = new ArrayList<>();
        int row = heights.length;
        int col = heights[0].length;
        if(row==0||col==0){
            return res;
        }
        boolean[][] ao = new boolean[row][col];
        boolean[][] po = new boolean[row][col];
        for(int i =0;i<col;i++){
            dfs(0,i,po,heights[0][i],heights);
            dfs(row-1,i,ao,heights[row-1][i],heights);
        }
        for(int j=0;j<row;j++){
            dfs(j,0,po,heights[j][0],heights);
            dfs(j,col-1,ao,heights[j][col-1],heights);
        }
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                if(po[i][j] && ao[i][j]){
                    List<Integer> temp = new ArrayList<>();
                    temp.add(i);
                    temp.add(j);
                    res.add(temp);
                }
            }
        }
        return res;
    }
    public void dfs(int r , int c , boolean [][] visited , int h , int[][]heights){
        if(r<0||c<0||r==heights.length||c==heights[0].length|| visited[r][c] || h>heights[r][c]){
            return;
        }
        visited[r][c] =true;
        dfs(r+1,c,visited,heights[r][c],heights);
        dfs(r,c+1,visited,heights[r][c],heights);
        dfs(r-1,c,visited,heights[r][c],heights);
        dfs(r,c-1,visited,heights[r][c],heights);
    }
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        n = len(heights)
        m = len(heights[0])
        pac = [[False] * m for _ in range(n)]
        alt = [[False] * m for _ in range(n)]
        directions = [(0,1),(1,0),(0,-1),(-1,0)]
        res = []
        def dfs(r,c,seen):
            seen[r][c] = True
            for dr,dc in directions:
                nr = r + dr
                nc = c + dc
                if 0 <= nr < n and 0 <= nc < m and heights[nr][nc] >= heights[r][c] and not seen[nr][nc]:
                    dfs(nr,nc,seen)
        for i in range(m):
            dfs(0,i,pac)
        for j in range(n):
            dfs(j,0,pac)
        for i in range(m):
            dfs(n-1,i,alt)
        for j in range(n):
            dfs(j,m-1,alt)
        for i in range(n):
            for j in range(m):
                if pac[i][j] and alt[i][j]:
                    res.append([i,j])
        return res
        